controllers/virtualmachine/volumebatch/volumebatch_controller.go

```diff
@@ -590,17 +598,90 @@ func (r *Reconciler) reconcileInstanceStoragePVCs(_ *pkgctx.VolumeContext) (bool
 // This ensures proper node selection for storage classes requiring it.
-func (r *Reconciler) handlePVCWithWFFC(_ *pkgctx.VolumeContext, _ vmopv1.VirtualMachineVolume) error {
-	// TODO: Implement WFFC (WaitForFirstConsumer) handling
-	// This method should:
-	// - Check storage class binding mode
-	// - Set selected node annotations for unbound PVCs
-	// - Handle node selection for proper PVC binding
-	// - Validate node availability and zone constraints
+func (r *Reconciler) HandlePVCWithWFFC(ctx *pkgctx.VolumeContext, volume vmopv1.VirtualMachineVolume) error {
+
+	if volume.PersistentVolumeClaim == nil || volume.PersistentVolumeClaim.InstanceVolumeClaim != nil {
+		return nil
+	}
+
+	pvc := corev1.PersistentVolumeClaim{}
+	pvcKey := client.ObjectKey{
+		Namespace: ctx.VM.Namespace,
+		Name:      volume.PersistentVolumeClaim.ClaimName,
+	}
+
+	if err := r.Get(ctx, pvcKey, &pvc); err != nil {
+		return fmt.Errorf("cannot get PVC: %w", err)
+	}
+
+	if pvc.Status.Phase == corev1.ClaimBound {
+		// Regardless of the StorageClass binding mode there is nothing to done if already bound.
+		return nil
+	}
+
+	if pvc.Annotations[constants.KubernetesSelectedNodeAnnotationKey] != "" {
+		// Try fetch the node object to check if it still exists. and the zone
+		// matches the VM's Zone.
+		node := &corev1.Node{}
+		nodeKey := client.ObjectKey{
+			Name: pvc.Annotations[constants.KubernetesSelectedNodeAnnotationKey],
+		}
+		if err := r.Get(ctx, nodeKey, node); err != nil {
+			return fmt.Errorf("cannot get Node %q: %w", nodeKey.Name, err)
+		}
+
+		if node.Labels[corev1.LabelTopologyZone] != ctx.VM.Status.Zone {
+			return fmt.Errorf("node %q is not in the VM's zone %q, but in zone %q",
+				node.Name, ctx.VM.Status.Zone, node.Labels[corev1.LabelTopologyZone])
+		}
+
+		return nil
+	}
+
+	scName := pvc.Spec.StorageClassName
+	if scName == nil {
+		return fmt.Errorf("PVC %s does not have StorageClassName set", pvc.Name)
+	} else if *scName == "" {
+		return nil
+	}
+
+	sc := &storagev1.StorageClass{}
+	if err := r.Get(ctx, client.ObjectKey{Name: *scName}, sc); err != nil {
+		return fmt.Errorf("cannot get StorageClass for PVC %s: %w", pvc.Name, err)
+	}
+
+	if mode := sc.VolumeBindingMode; mode == nil ||
+		*mode != storagev1.VolumeBindingWaitForFirstConsumer {
+		return nil
+	}
+
+	if !pkgcfg.FromContext(ctx).Features.VMWaitForFirstConsumerPVC {
+		return fmt.Errorf("PVC with WFFC storage class support is not enabled")
+	}
+
+	zoneName := ctx.VM.Status.Zone
+	if zoneName == "" {
+		// Fallback to the label value if Status hasn't been updated yet.
+		zoneName = ctx.VM.Labels[corev1.LabelTopologyZone]
+		if zoneName == "" {
+			return fmt.Errorf("VM does not have Zone set")
+		}
+	}
+
+	if pvc.Annotations == nil {
+		pvc.Annotations = map[string]string{}
+	}
+	pvc.Annotations[volumebatchutils.CNSSelectedNodeIsZoneAnnotationKey] = "true"
+	pvc.Annotations[constants.KubernetesSelectedNodeAnnotationKey] = zoneName
```

Comment:
Nit: Consider using this const from upstream: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-helpers/storage/volume/pv_helpers.go#L51